### ISSUE TITLE
Dynamo can not optimize a model with MaxPool2d on XLA devices

### ISSUE THREAD

**Issue by shunting314** [pytorch/torchdynamo#1837]
### 🐛 Describe the bug

Found this bug when integrate dynamo with torchxla for resnet model. If we move the model and its inputs to XLA device before running dynamo, we would hit this bug. Check the minimal repro below.

cc @jansel @wconstab @jackcaog

### Error logs

  File "/pytorch/torch/_dynamo/convert_frame.py", line 118, in _fn
    return fn(*args, **kwargs)
  File "/pytorch/torch/_dynamo/utils.py", line 92, in time_wrapper
  File "/pytorch/torch/_dynamo/convert_frame.py", line 118, in _fn
  File "/pytorch/torch/_refs/__init__.py", line 45, in <module>
    from torch.fx.experimental.symbolic_shapes import sym_float, sym_int
  File "/pytorch/torch/fx/experimental/symbolic_shapes.py", line 17, in <module>
    import sympy  # type: ignore[import]
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/__init__.py", line 51, in <module>
    from .core import (sympify, SympifyError, cacheit, Basic, Atom,
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/core/__init__.py", line 4, in <module>
    from .sympify import sympify, SympifyError
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/core/sympify.py", line 9, in <module>
    from .compatibility import iterable
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/core/compatibility.py", line 11, in <module>
    from sympy.external import import_module
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/external/__init__.py", line 18, in <module>
    from sympy.external.importtools import import_module
  File "/root/anaconda3/envs/pytorch/lib/python3.7/site-packages/sympy/external/importtools.py", line 4, in <module>
    from distutils.version import LooseVersion
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 963, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 906, in _find_spec
    return fn(*args, **kwargs)
  File "/pytorch/torch/_dynamo/utils.py", line 92, in time_wrapper
    r = func(*args, **kwargs)
  File "/pytorch/torch/_dynamo/convert_frame.py", line 356, in _convert_frame_assert
    frame,
  File "/pytorch/torch/_dynamo/convert_frame.py", line 402, in _compile
    out_code = transform_code_object(code, transform)
  File "/pytorch/torch/_dynamo/bytecode_transformation.py", line 341, in transform_code_object
    transformations(instructions, code_options)
  File "/pytorch/torch/_dynamo/convert_frame.py", line 390, in transform
    tracer.run()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 1468, in run
    super().run()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 352, in run
    and self.step()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 322, in step
    getattr(self, inst.opname)(inst)
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 174, in wrapper
    return inner_fn(self, inst)
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 766, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 264, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/pytorch/torch/_dynamo/variables/nn_module.py", line 209, in call_function
    **options,
  File "/pytorch/torch/_dynamo/convert_frame.py", line 118, in _fn
    return fn(*args, **kwargs)
  File "/pytorch/torch/_dynamo/utils.py", line 92, in time_wrapper
    r = func(*args, **kwargs)
  File "/pytorch/torch/_dynamo/convert_frame.py", line 356, in _convert_frame_assert
    frame,
  File "/pytorch/torch/_dynamo/convert_frame.py", line 402, in _compile
    out_code = transform_code_object(code, transform)
  File "/pytorch/torch/_dynamo/bytecode_transformation.py", line 341, in transform_code_object
    transformations(instructions, code_options)
  File "/pytorch/torch/_dynamo/convert_frame.py", line 390, in transform
    tracer.run()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 1468, in run
    super().run()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 352, in run
    and self.step()
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 322, in step
    getattr(self, inst.opname)(inst)
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 174, in wrapper
    return inner_fn(self, inst)
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 766, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/pytorch/torch/_dynamo/symbolic_convert.py", line 264, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/pytorch/torch/_dynamo/variables/nn_module.py", line 209, in call_function
    **options,
  File "/pytorch/torch/_dynamo/variables/tensor.py", line 201, in create
    example_value = _get_fake_value(proxy.node, tx)
  File "/pytorch/torch/_dynamo/variables/tensor.py", line 145, in _get_fake_value
    raise TorchRuntimeError() from e
torch._dynamo.exc.TorchRuntimeError:

from user code:
   File "myscripts/repro_maxpool.py", line 14, in forward
    out = self.pool(out)

Set torch._dynamo.config.verbose=True for more information

You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True

### Minified repro

repro_maxpool.py
```
from torch import nn
import torch
import torch._dynamo as dynamo
import torch_xla.core.xla_model as xm

class MaxPoolModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 6, kernel_size=3, stride=2)
        self.pool = nn.MaxPool2d(kernel_size=3, stride=2)

    def forward(self, x):
        out = self.conv(x)
        out = self.pool(out)
        return out

    def get_random_inputs(self):
        return (torch.rand(2, 3, 10, 10),)

xla_dev = xm.xla_device()
model = MaxPoolModule().to(device=xla_dev)
inputs = map(lambda x: x.to(device=xla_dev), model.get_random_inputs())
dynamo.optimize(lambda gm, _: gm)(lambda: model(*inputs))()
```

Command:
```
GPU_NUM_DEVICES=1 python repro_maxpool.py
```

**Comment by JackCaoG** [comment]
maxpool2d is a bit special, it is one of those autograd op that pytorch/xla manually register the backward hook in https://github.com/pytorch/xla/blob/master/xla_native_functions.yaml#L357. It seems like dynamo can not handle this kind of use case? 

@bdhirsh FYI

**Comment by shunting314** [comment]
MaxPool2d is used in so many vision models that this issue will affect at least the following models I've tested:
```
resnet18, resnet50, resnext50_32x4d, alexnet, vgg16, squeezenet1_1
```

**Comment by JackCaoG** [comment]
hmm if you leave the model in cpu this issue won't happen? That's a bit odd.

My only theory is if the model is on cpu, dynamo won't know about xla's special handling of `max_pool2d`. However if that's the case, in the backward dynamo would ignore the special backward of xla in [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_xla_type.cpp#L1560) which calls [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_autograd_ops.h#L28)

**Comment by shunting314** [comment]
> hmm if you leave the model in cpu this issue won't happen? That's a bit odd.

Right, the issue only happens if the model is moved to XLA devices before running dynamo

**Comment by shunting314** [comment]
I actually don't think the issue is related to backward pass since disabling that can also repro the issue:
```
@torch.no_grad()
class MaxPoolModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 6, kernel_size=3, stride=2)
        self.pool = nn.MaxPool2d(kernel_size=3, stride=2)

    def forward(self, x):
        out = self.conv(x)
        out = self.pool(out)
        return out

    def get_random_inputs(self):
        return (torch.rand(2, 3, 10, 10),)
```

**Comment by JackCaoG** [comment]
Ok I guess the issue is that op under `autograd` category (which let backend to implement custom backward) can not be recognized by the dynamo. To confirm this theory, can you try `max_pool3d`, `native_layer_norm` and `einsum` and see if you run into the same issue?

**Comment by wconstab** [comment]
I see the stack trace failing in symbolic tracing, when we try to run a fake tensor through an operation to get its fake output as a part of building the initial FX forward graph inside dynamo.  This is before we use aot-autograd or call backwards.

The bottom of the stacktrace shows us hitting the failing call to run_node ([here - tensor.py#L132](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/tensor.py#L132) and re-raising [here - tensor.py#L145](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/tensor.py#L145).  Note there should be 2 stacktraces.  one came from 'e' from L132, and the other came from reraise L145.  If you look at the top of your file, i think the first (real) stacktrace is above the second one, squished together in a way you can't read them.  it looks like you have issues importing sympy..   maybe this is just happening bc your version of maxpool actually returned dynamic shapes?

**Comment by bdhirsh** [comment]
Agreed with Will - it's a bit hard to tell what the real error from FakeTensor is.

Jack is correct that `max_pool2d` is somewhat of a special case for XLA. In PyTorch core, we have a decomposition directly for `max_pool2d`. In PyTorch/XLA, that decomposition gets overridden and replaced with a custom lowering for the forward backward.

Given that the error is around fake tensor, my guess is that FakeTensor isn't able to run a meta function for `max_pool2d`, since the decomposition is getting ignored / overridden in XLA.

One thing you can try is manually registering a meta function. If this fixes the issue, we can think about doing this in XLA in a more principled way:
```
from torch._meta_registrations import register_meta

@register_meta(torch.ops.aten.max_pool2d.default)
def max_pool2d_meta(self, kernel_size, stride=[], padding=0, dilation=1, ceil_mode=False):
    return torch.ops.aten.max_pool2d.default.decompose(self, kernel_size, stride, padding, dilation, ceil_mode)
```

**Comment by shunting314** [comment]
> If you look at the top of your file, i think the first (real) stacktrace is above the second one, squished together in a way you can't read them.

@wconstab  check below for the original exception. Look like somewhere expect a XLA tensor but get something else. 

```
Traceback (most recent call last):
  File "/pytorch/torch/_dynamo/variables/tensor.py", line 132, in _get_fake_value
    lambda: _run_node(tx.output, node, args, kwargs, nnmodule)
  File "/pytorch/torch/_dynamo/utils.py", line 732, in wrap_fake_exception
    return fn()
  File "/pytorch/torch/_dynamo/variables/tensor.py", line 132, in <lambda>
    lambda: _run_node(tx.output, node, args, kwargs, nnmodule)
  File "/pytorch/torch/_dynamo/variables/tensor.py", line 56, in _run_node
    return nnmodule(*args, **kwargs)
  File "/pytorch/torch/nn/modules/module.py", line 1423, in _call_impl
    return forward_call(*input, **kwargs)
  File "/pytorch/torch/nn/modules/pooling.py", line 168, in forward
    return_indices=self.return_indices)
  File "/pytorch/torch/_jit_internal.py", line 485, in fn
    return if_false(*args, **kwargs)
  File "/pytorch/torch/nn/functional.py", line 782, in _max_pool2d
    return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:73 : Check failed: xtensor
```

**Comment by shunting314** [comment]
> One thing you can try is manually registering a meta function. If this fixes the issue, we can think about doing this in XLA in a more principled way:

@bdhirsh  tried this, still the same error

**Comment by shunting314** [comment]
Here is where everything inputs looks like before calling torch.max_pool2d:
```
> /pytorch/torch/nn/functional.py(784)_max_pool2d()
-> ret = torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
(Pdb) p input
FakeTensor(FakeTensor(..., device='meta', size=(2, 6, 4, 4),
           grad_fn=<ConvolutionBackward0>), xla:1)
(Pdb) p kernel_size
3
(Pdb) p stride
2
(Pdb) p type(kernel_size)
<class 'int'>
(Pdb) p padding
0
(Pdb) p dilation
1
(Pdb) p ceil_mode
False
```

**Comment by bdhirsh** [comment]
@shunting314 just as a test, can you give this patch a try (in pytorch/XLA)? I posted my theory below.
```
--- a/torch_xla/csrc/aten_autograd_ops.cpp
+++ b/torch_xla/csrc/aten_autograd_ops.cpp
@@ -2,6 +2,7 @@

 #include <ATen/Operators.h>
 #include <ATen/native/CPUFallback.h>
+#include <ATen/RedispatchFunctions.h>

 #include "torch_xla/csrc/aten_cpu_fallback.h"
 #include "torch_xla/csrc/aten_xla_bridge.h"
@@ -71,6 +72,15 @@ torch::Tensor MaxPool2dAutogradFunction::forward(
   ctx->saved_data["dilation"] = dilation;
   ctx->saved_data["ceil_mode"] = ceil_mode;
   // Lowering when ceil_mode or dilation is set not supported yet.
+  if (self.key_set().has(c10::DispatchKey::Python) {
+    // We might get here after the autograd graph was created, but before
+    // python gets a chance to run. Give python a chance to run.
+    return at::redispatch::max_pool2d(
+      c10::DispatchKeySet(c10::DispatchKey::Python),
+      self, kernel_size, stride, padding, dilation, ceil_mode);
+  }
+
+  }
   if (IsNonTrivialDilation(dilation)) {
```

> RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:73 : Check failed: xtensor

That error really looks like fake tensor is trying to directly call the XLA lowering for `max_pool2d` (and it can't because it's passing in `FakeTensor` inputs, and not real XLA tensors).

Oh that's pretty interesting. I think what's happening is the following:

In the "normal" case for autograd, for some op like `foo`, we get autograd behavior through two dispatch keys:

`DispatchKey::AutogradCPU`: the function registered to this key doesn't do any real compute: it sets up some nodes in the autograd graph (`foo_backward()`, and redispatches to the actual backend to call the kernel for the forward function

`DispatchKey::CPU`: The actual kernel for `foo` is registered here.


The relative ordering of the dispatch keys that are relevant here is:
```
AutogradCPU
Python (lower priority than autograd, higher priority than other backends)
CPU
```

What happens with the fake tensor tracing in dynamo is that when you call an op like `torch.foo()` with fake (cpu) tensors, we run:
```
AutogradCPU key: doesn't run real compute, adds `foo_backward` to the autograd graph
Python key: we end up getting here BEFORE hitting the CPU key, which takes us straight to to `FakeTensorMode`. The actual CPU kernel never ends up running
```

Now, this setup will do something interesting with XLA's treatment of `max_pool2d`. Today, XLA lowers `max_pool2d` by registering a custom `autograd::Function` object, **directly** to the `AutogradXLA` dispatch key. So when torchdynamo runs `torch.max_pool2d` with fake tensors, the ordering goes:

```
(1) AutogradXLA key -> run the custom autograd::Function::apply() object that XLA registered
(2) (hopefully?) Python key
```

But the Python key doesn't get a chance to run - the custom autograd::function() object will (a) set up the backward graph, and (b) call you forward function, all in one go! So we pass fake tensors to the forward function for XLA, at which point we error.

cc @albanD + @zou3519. Still need to confirm this theory, but seems like a case for a "better custom autograd function"?

**Comment by shunting314** [comment]
@bdhirsh  that magically works (after fixing two typos :) ).

To understand more, even though the issue also happens in inference scenario, the fundamental reason is torchxla need customize the autograd behavior of max_pool2d. So it has a autograd::Function for max_pool2d. And the forward part of that autograd::Function can not handle fake tensor properly. Correct?

Also do you want to send a PR for this?

**Comment by bdhirsh** [comment]
Here's a patch, although we could consider a different fix longer term: https://github.com/pytorch/xla/pull/4165

**Comment by JackCaoG** [comment]
https://github.com/pytorch/xla/pull/4165 is merged, do we want to keep this one open to track the long term fix?

**Comment by zou3519** [comment]
> The problem is that "setting up the autograd graph" and "running the forward kernel" are normally split across 2 dispatch keys. With autograd::function::apply() though, they're fused together (so they both run as part of the autograd dispatch key).

Yes, that is one of the motivations for a better autograd.Function

**Comment by shunting314** [comment]
PR https://github.com/pytorch/xla/pull/4165 seems to only fix the forward case. When enabling `aot_autograd`, we see failure again.

Here is the repro for training:
```
from torch import nn
import torch
import torch._dynamo as dynamo
import torch_xla.core.xla_model as xm
import os
from functorch.compile import aot_module_simplified
from torch._dynamo import disable
from functorch.compile import make_boxed_compiler
os.environ["GPU_NUM_DEVICES"] = "1"

class MaxPoolModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv2d(3, 6, kernel_size=3, stride=2)
        self.pool = nn.MaxPool2d(kernel_size=3, stride=2)

    def forward(self, x):
        out = self.conv(x)
        out = self.pool(out)
        return out

    def get_random_inputs(self):
        return (torch.rand(2, 3, 10, 10),)

xla_dev = xm.xla_device()
model = MaxPoolModule().to(device=xla_dev)
inputs = map(lambda x: x.to(device=xla_dev), model.get_random_inputs())

def compiler(gm, inputs):
    @make_boxed_compiler
    @disable
    def fw_compiler(graph, inputs, *args, **kwargs):
        return graph
    return aot_module_simplified(gm, fw_compiler=fw_compiler)

dynamo.optimize(compiler)(lambda: model(*inputs).sum().backward())()

print("bye")
```

Stack trace is too long, here is part of it:
```
  File "/pytorch/torch/fx/interpreter.py", line 171, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/pytorch/torch/fx/interpreter.py", line 288, in call_module
    return submod(*args, **kwargs)
  File "/pytorch/torch/fx/_symbolic_trace.py", line 717, in module_call_wrapper
    return self.call_module(mod, forward, args, kwargs)
  File "/pytorch/torch/fx/experimental/proxy_tensor.py", line 378, in call_module
    return forward(*args, **kwargs)
  File "/pytorch/torch/fx/_symbolic_trace.py", line 710, in forward
    return _orig_module_call(mod, *args, **kwargs)
  File "/pytorch/torch/nn/modules/module.py", line 1423, in _call_impl
    return forward_call(*input, **kwargs)
  File "/pytorch/torch/nn/modules/pooling.py", line 168, in forward
    return_indices=self.return_indices)
  File "/pytorch/torch/_jit_internal.py", line 485, in fn
    return if_false(*args, **kwargs)
  File "/pytorch/torch/nn/functional.py", line 782, in _max_pool2d
    return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
RuntimeError: /pytorch/xla/torch_xla/csrc/aten_xla_bridge.cpp:73 : Check failed: xtensor
```

cc @JackCaoG  @wconstab @bdhirsh

**Comment by shunting314** [comment]
This part of log may also be relevant:
```
*** End stack trace ***
Input tensor is not an XLA tensor: XLAFloatType
```

It's weird that it says a XLAFloatType is not an XLA tensor.

**Comment by shunting314** [comment]
Full log can be found here: https://gist.github.com/shunting314/da43fe460d86b24b97027bb6c34efd7b

**Comment by shunting314** [comment]
Here is a summary of this issue. When doing dynamo/torchxla integration, we found models using maxpool2d with fail for inference. Check ( https://github.com/pytorch/torchdynamo/issues/1837#issue-1435394861 ) for description and repro. Brian comes up with a workaround ( https://github.com/pytorch/xla/pull/4165 ) and unblocks inference for those models.  However we found there is still issues about maxpool2d for training. Check ( https://github.com/pytorch/torchdynamo/issues/1837#issuecomment-1312287846 ) for repro for training.

cc @JackCaoG @wconstab @ezyang

**Comment by shunting314** [comment]
Some updates:

I found when crash happens the input to `torch.max_pool2d` is a FunctionalTensor
```
_to_functional_tensor(tensor([[[[ 0.0924,  0.1429,  0.0210,  0.2755],
          [ 0.1148,  0.0290, -0.0287, -0.1240],
          [ 0.4501, -0.0103,  0.3415, -0.0189],
          [-0.0530, -0.1287, -0.2060, -0.1524]],
```

so I disabled functionalization in functorch in the repro and now the repro works. But maybe we should not disable functionalization in general because AOTAutograd needed that? cc @bdhirsh

**Comment by shunting314** [comment]
Confirmed functionalize can not be disabled in general. When disabling functionaliztion to train resnet, I get the following woarning:
```
[WARNING] Unable to use Aot Autograd because of presence of mutation
```

**Comment by wconstab** [comment]
So, I think @bdhirsh will be able to help with an alternative real solution for max pool and other custom autograd ops, but he's doing a bunch of things this week.

Just to try to decouple this issue from the broader question of getting training working, can we just change the model to avoid this op and get an e2e training without maxpool working first?  Either by picking a model that doesn't use maxpool, or even by changing maxpool to like.. a slice with the same shape as maxpool output?

**Comment by bdhirsh** [comment]
Oh- I’m not at my laptop, but this should be fixable.

instead of redispatching directly to the python key, we need to make sure that we can hit any other dispatch keys that need to run first, like functionalization.

**Comment by shunting314** [comment]
@JackCaoG , @wconstab  any idea why AvgPool2d works (at least in the repro I provided) , while MaxPool2d fail . What's the difference how XLA handle them?

**Comment by shunting314** [comment]
Training for resnet18 works now with aot_eager or aot_torchxla_trivial backend if replacing MaxPool to AvgPool. `aot_torchxla_trace_once` still have some problem, but that's not related to this issue. Overall, replace MaxPool to AvgPool can temporarily unblock the project before the real fix is in.
